test-data/unit/check-parameter-specification.test

```diff
@@ -406,3 +405,224 @@ with f() as x:
+T = TypeVar("T")
+P_2 = ParamSpec("P_2")
+
+class X(Generic[T, P]):
+  f: Callable[P, int]
+  x: T
+
+def f1(x: X[int, P_2]) -> str: ...                    # Accepted
+def f2(x: X[int, Concatenate[int, P_2]]) -> str: ...  # Accepted
+def f3(x: X[int, [int, bool]]) -> str: ...            # Accepted
+# Is ellipsis allowed by PEP? This shows up:
+# def f4(x: X[int, ...]) -> str: ...                    # Accepted
+# TODO: this is not rejected:
+# def f5(x: X[int, int]) -> str: ...                    # Rejected
+
+# CASE 3
+def bar(x: int, *args: bool) -> int: ...
+def add(x: Callable[P, int]) -> Callable[Concatenate[str, P], bool]: ...
+
+reveal_type(add(bar))       # N: Revealed type is "def (builtins.str, x: builtins.int, *args: builtins.bool) -> builtins.bool"
+
+def remove(x: Callable[Concatenate[int, P], int]) -> Callable[P, bool]: ...
+
+reveal_type(remove(bar))    # N: Revealed type is "def (*args: builtins.bool) -> builtins.bool"
+
+def transform(
+  x: Callable[Concatenate[int, P], int]
+) -> Callable[Concatenate[str, P], bool]: ...
+
+# In the PEP, "__a" appears. What is that? Autogenerated names? To what spec?
+reveal_type(transform(bar)) # N: Revealed type is "def (builtins.str, *args: builtins.bool) -> builtins.bool"
+
+# CASE 4
+def expects_int_first(x: Callable[Concatenate[int, P], int]) -> None: ...
+
+@expects_int_first  # E: Argument 1 to "expects_int_first" has incompatible type "Callable[[str], int]"; expected "Callable[[int], int]"
+def one(x: str) -> int: ...
+
+@expects_int_first  # E: Argument 1 to "expects_int_first" has incompatible type "Callable[[NamedArg(int, 'x')], int]"; expected "Callable[[int], int]"
+def two(*, x: int) -> int: ...
+
+@expects_int_first  # E: Argument 1 to "expects_int_first" has incompatible type "Callable[[KwArg(int)], int]"; expected "Callable[[int], int]"
+def three(**kwargs: int) -> int: ...
+
+@expects_int_first # Accepted
+def four(*args: int) -> int: ...
+[builtins fixtures/tuple.pyi]
+[builtins fixtures/dict.pyi]
+
+[case testParamSpecTwiceSolving]
+from typing_extensions import ParamSpec, Concatenate
+from typing import Callable, TypeVar
+
+P = ParamSpec("P")
+R = TypeVar("R")
+
+def f(one: Callable[Concatenate[int, P], R], two: Callable[Concatenate[str, P], R]) -> Callable[P, R]: ...
+
+a: Callable[[int, bytes], str]
+b: Callable[[str, bytes], str]
+
+reveal_type(f(a, b))  # N: Revealed type is "def (builtins.bytes) -> builtins.str*"
+[builtins fixtures/tuple.pyi]
+
+[case testParamSpecConcatenateInReturn]
+from typing_extensions import ParamSpec, Concatenate
+from typing import Callable, Protocol
+
+P = ParamSpec("P")
+
+def f(i: Callable[Concatenate[int, P], str]) -> Callable[Concatenate[int, P], str]: ...
+
+n: Callable[[int, bytes], str]
+
+reveal_type(f(n))  # N: Revealed type is "def (builtins.int, builtins.bytes) -> builtins.str"
+[builtins fixtures/tuple.pyi]
+
+[case testParamSpecConcatenateNamedArgs]
+# this is one noticeable deviation from PEP but I believe it is for the better
```

Comment:
I'm no longer sure about this behavior. I did this in spirit of https://github.com/python/mypy/issues/11606

Reply:
Clarification:

The PEP simultaneously says (emphasis mine):
> The semantics of `Concatenate[X, Y, P]` are that it represents the parameters represented by `P` with two **positional-only** parameters prepended.

and also gives this example:
```py
def add(f: Callable[P, int]) -> Callable[Concatenate[str, P], None]:

  def foo(s: str, *args: P.args, **kwargs: P.kwargs) -> None:  # Accepted
    pass

  def bar(*args: P.args, s: str, **kwargs: P.kwargs) -> None:  # Rejected
    pass

  return foo                                                   # Accepted
```
  
Which is... a confusing message to be giving off, to say the least.

Reply:
It indeed looks looks the PEP authors forgot to adjust the runtime semantics, i.e. requiring a `/` between `x` and `*args`. The example from the PEP
```py
def outer(f: Callable[P, None]) -> Callable[P, None]:
  def foo(x: int, *args: P.args, **kwargs: P.kwargs) -> None:
    f(*args, **kwargs)

  def bar(*args: P.args, **kwargs: P.kwargs) -> None:
    foo(1, *args, **kwargs)   # Accepted
    foo(x=1, *args, **kwargs) # Rejected

  return bar
```

```py
@outer
def problem(*, x: object) -> None:
  pass

problem(x="uh-oh")
```

Even the "accepted" case raises a runtime error. This is probably something worth discussing over at [python/typing](https://github.com/python/typing/issues?page=1&q=is%3Aissue+is%3Aopen+label%3A%22topic%3A+feature%22).

Reply:
Added a strictness flag.